ir/type.cpp

```diff
@@ -231,10 +236,47 @@ void VectorType::print(ostream &os) const {
   os << "TODO";
 }
 
+expr AggregateType::getTypeConstraints() const {
+  // Only allow aggregates of type {ix, i1} for overflow instrs
+  return childrenSize.size() == 2 &&
+         childrenSize[1] == 1;
+}
+
+expr AggregateType::enforceAggregateType() const {
+  return true;
+}
+
+void AggregateType::fixup(const Model &m) {
+  // TODO
+}
+
+expr AggregateType::enforceIntOrPtrOrVectorType() const {
+  return false;
+}
+
+unsigned AggregateType::bits() const {
+  return childrenSize[0] + childrenSize[1];
+}
+
+void AggregateType::print(ostream &os) const {
+  assert(childrenSize.size() > 0);
+  os << "{";
+  os << "i" << childrenSize[0];
+  for (unsigned i = 1; i < childrenSize.size(); i++) {
+    os << ", i" << childrenSize[i];
+  }
+  os << "}";
+}
+
+expr AggregateType::getChildrenConstraints(Type &type) const {
+  return (type.enforceIntType() && type.sizeVar() == childrenSize[0]) ||
+         (type.enforceIntType() && type.sizeVar() == childrenSize[1]);
```

Comment:
can you generalize this so that aggregate type is a container of arbitrary types pls?

Reply:
do you want to make use of Z3 records/tuple for this purpose? Currently, I am using bitvectors as representation of aggregate type (since we currently only allow integers as children components and it's enough for overflow intrinsics). Not sure, but adding support for arbitrary types for aggregate type with aribtrary number of components may not be possible with just bitvectors.

Reply:
@pranavk I don't think it makes sense that something like this isn't possible using bitvectors -- alive2 simple has to keep track of the type information instead of relying on z3

@nunoplopes I had mentioned to Pranav that you said that Z3 data types don't play well with quantifiers -- I assume that this forces a low-level solution here?

Reply:
Right, no Z3 datatypes.
You can continue to use bitvectors. The result will be a concatenation of the underlying representation of each of the components (which will all be bitvectors).